### ISSUE TITLE
Bug: [no-unnecessary-boolean-literal-compare] should require `strictNullChecks `

### ISSUE THREAD

**Issue by earshinov** [typescript-eslint/typescript-eslint#10706]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.ts&code=GYVwdgxgLglg9mABFApgZygCgIYH4BciARnHADYrZgCUhJ5lSA3gLABQiiATilCF0myIAhAF5RiYNjJoUAbnYBfdu1QZM1OUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ARhw5IhTWo3yJoQ%2BLTIcAtsRGJ0UJdA7RI4MAF8QZoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYAZgIaonrhgC%2BIfkA&tokens=false

### Repro Code

```TypeScript
function test(a?: boolean): boolean {
  return a !== false;
}

test();
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": false
  }
}
```

### Actual Result

With the configuration above (without `strictNullChecks`), the `@typescript-eslint/no-unnecessary-boolean-literal-compare` rule emits a (false) error. Applying the automatic fix translates

```typescript
function test(a?: boolean): boolean {
  return a !== false;
}
```

into

```typescript
function test(a?: boolean): boolean {
  return a;
}
```

which is obviously not equivalent as `a` is nullable.

Root cause seems to be that without `strictNullChecks`, TypeScript reports the type of `a` as `boolean` instead of `boolean | undefined`, making typescript-eslint think that `a` is not nullable ⇒ `a !== false` is equivalent to just `a`.

My guess is that the `no-unnecessary-boolean-literal-compare` rule absolutely has to check for `strictNullChecks` and refuse to work (like other rules do, such as `prefer-nullish-coalescing`).

We already broke our app in runtime after applying `eslint --fix` without reviewing each and every change. We can't enable `strict` / `strictNullChecks` as it would emit thousands of errors in the legacy codebase. 

### Expected Result

Either of
a. (best option) no error, like with `strictNullChecks: true` (I assume, not technically possible)
b. eslint refuses to work if the `no-unnecessary-boolean-literal-compare` rule is enabled, but `strictNullChecks` aren't

**Comment by kirkwaiblinger** [comment]
Hi @earshinov!

I think it's reasonable for us to treat this rule the same way as others that require strict null checks, like [strict-boolean-expressions](https://typescript-eslint.io/rules/strict-boolean-expressions/#allowruletorunwithoutstrictnullchecksiknowwhatiamdoing).

Personally, I'm also wary of subtly runtime-behavior-changing autofixes that rely on type information with a reasonable likelihood of being wrong (for similar issues see #10136 and its predecessors). So I'd be in favor of downgrading the fixes to suggestions for this rule.

**Comment by bradzacher** [comment]
+1 I think that we should treat this the same and break it without `strictNullChecks` like the other rules. It doesn't make sense without it.
